astroid/inference.py

```diff
@@ -89,6 +89,23 @@ def infer_map(self, context=None):
         yield new_seq
 
 
+def _update_with_replacement(inparam_dict, new_dict):
+    """Delete nodes that equate to duplicate keys
+
+    Fixes issue with DictUnpack causing duplicte keys
+    in inferred Dict items
+
+    Modifies inparam_dict
+    """
+    new_dict_key_nodes = tuple(new_dict)
```

Comment:
I think the algorithm could be simplified a bit.

We know that the last value wins in a dictionary, so `{**data, **some_other_data, 'key': 1}` will have the last `key` set even if `data` or `some_other_data` would contain the same key name. What we could do is to hash the node and store the hash inside the `values` dictionary instead of storing the node itself (which can be stored in a separate map for reconstruction). Then, because keys having the same name will have the same hash, the last one will win, since we keep a direct iteration through `node.items`. The only case that needs to be handled separately though is that duplicate strings from non-`DictUnpack` nodes need to still be present in the dictionary, since we use that for a pylint check, as I mentioned in another comment. With this suggestion though, we don't need a separate `contents_equal` method and the algorithm could just be in-place in the `_infer_map` function.

Reply:
I think I see what you mean... I can make this function O(n) instead of O(n^2) if I use some kind of hash. I'm unsure how to "hash" the node however. The key node could be anything from a const, to a tuple, to a custom class with a __hash__ function.

Reply:
I guess I could infer each key and hash the `as_string()`?

Reply:
Sure, `as_string` sounds like a good hash solution to me. It will result in same values for potentially different objects, such as `**d, **d` but that is exactly what we want in this case